File: app/src/main/java/de/rwth_aachen/phyphox/ExperimentList/handler/BluetoothScanner.java
package de.rwth_aachen.phyphox.ExperimentList.handler;

import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.content.res.Resources;
import android.os.AsyncTask;
import android.os.Build;

import java.util.Set;
import java.util.UUID;

import de.rwth_aachen.phyphox.Bluetooth.Bluetooth;
import de.rwth_aachen.phyphox.Bluetooth.BluetoothScanDialog;
import de.rwth_aachen.phyphox.R;

//The BluetoothScanDialog has been written to block execution until a device is found, so we should not run it on the UI thread.
public class BluetoothScanner extends AsyncTask<String, Void, BluetoothScanDialog.BluetoothDeviceInfo> {

    public interface BluetoothScanListener {
        void onBluetoothDeviceFound(BluetoothScanDialog.BluetoothDeviceInfo result);

        void onBluetoothScanError(String msg, Boolean isError, Boolean isFatal);
    }

    private final BluetoothScanner.BluetoothScanListener listener;
    private final Resources res;

    private Activity parent;

    Set<String> bluetoothDeviceNameList;
    Set<UUID> bluetoothDeviceUUIDList;

    public BluetoothScanner(Activity parent, Set<String> bluetoothDeviceNameList, Set<UUID> bluetoothDeviceUUIDList, BluetoothScanner.BluetoothScanListener listener) {
        this.listener = listener;
        this.parent = parent;
        this.res = parent.getResources();
        this.bluetoothDeviceNameList = bluetoothDeviceNameList;
        this.bluetoothDeviceUUIDList = bluetoothDeviceUUIDList;

    }

    //Copying is done on a second thread...
    protected BluetoothScanDialog.BluetoothDeviceInfo doInBackground(String... params) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2 || !Bluetooth.isSupported(parent)) {
            listener.onBluetoothScanError(res.getString(R.string.bt_android_version), true, true);
            return null;
        } else {
            BluetoothAdapter btAdapter = BluetoothAdapter.getDefaultAdapter();
            if (btAdapter == null || !Bluetooth.isEnabled()) {
                listener.onBluetoothScanError(res.getString(R.string.bt_exception_disabled), true, false);
                return null;
            }
            BluetoothScanDialog bsd = new BluetoothScanDialog(false, parent, parent, btAdapter);

            return bsd.getBluetoothDevice(null, null, bluetoothDeviceNameList, bluetoothDeviceUUIDList, null);
        }
    }

    @Override
    protected void onPostExecute(BluetoothScanDialog.BluetoothDeviceInfo result) {
        if (result != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2)
            listener.onBluetoothDeviceFound(result);
    }
}

File: app/src/main/java/de/rwth_aachen/phyphox/ExperimentList/handler/BluetoothScanner.kt
package de.rwth_aachen.phyphox.ExperimentList.handler

import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.os.Handler
import android.os.Looper
import de.rwth_aachen.phyphox.Bluetooth.Bluetooth
import de.rwth_aachen.phyphox.Bluetooth.BluetoothScanDialog
import de.rwth_aachen.phyphox.R
import java.util.UUID

/**
 * Scans from the experiment list (no experiment loaded) for devices matching a bundled experiment
 * or advertising the phyphox service. The blocking [BluetoothScanDialog] runs on a background
 * thread; the listener is called on the main thread.
 */
class BluetoothScanner(
    private val parent: Activity,
    private val bluetoothDeviceNameList: Set<String>,
    private val bluetoothDeviceUUIDList: Set<UUID>,
    private val listener: BluetoothScanListener
) {

    interface BluetoothScanListener {
        fun onBluetoothDeviceFound(result: BluetoothScanDialog.BluetoothDeviceInfo)
        fun onBluetoothScanError(msg: String, isError: Boolean, isFatal: Boolean)
    }

    fun execute() {
        val mainHandler = Handler(Looper.getMainLooper())
        Thread {
            val res = parent.resources
            if (!Bluetooth.isSupported(parent)) {
                mainHandler.post { listener.onBluetoothScanError(res.getString(R.string.bt_android_version), true, true) }
            } else if (BluetoothAdapter.getDefaultAdapter() == null || !Bluetooth.isEnabled()) {
                mainHandler.post { listener.onBluetoothScanError(res.getString(R.string.bt_exception_disabled), true, false) }
            } else {
                val scanDialog = BluetoothScanDialog(false, parent, parent, BluetoothAdapter.getDefaultAdapter())
                val result = scanDialog.getBluetoothDevice(null, null, bluetoothDeviceNameList, bluetoothDeviceUUIDList, null)
                val failure = scanDialog.scanFailureCode
                if (result != null)
                    mainHandler.post { listener.onBluetoothDeviceFound(result) }
                else if (failure != null) {
                    val msg = res.getString(R.string.bt_scan_failed) + " " + failure
                    mainHandler.post { listener.onBluetoothScanError(msg, true, false) }
                }
            }
        }.start()
    }
}
